### PR TITLE
Self staking dev

### PR DESCRIPTION
**Pull request by DeV1doR**

### CODE REVIEW

File: action/staking/stake.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		return false, errors.Wrap(action.ErrInvalidAmount, coin.String())
+	}
+
+	if coin.Currency.Name != "OLT" {
+		return false, action.ErrInvalidCurrency
+	}
+
+	_, ok := ctx.Currencies.GetCurrencyByName("OLT")
+	if !ok {
+		return false, action.ErrInvalidCurrency
```
Comment (ws4charlie) [inline action/staking/stake.go:100]: Is it better to panic here?

File: data/delegation/store.go [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+	"github.com/Oneledger/protocol/data/keys"
+	"github.com/Oneledger/protocol/serialize"
+	"github.com/Oneledger/protocol/storage"
+)
+
+type DelegationStore struct {
+	state  *storage.State
+	szlr   serialize.Serializer
+	prefix []byte
+	mux    sync.Mutex
```
Comment (ws4charlie) [inline data/delegation/store.go:17]: Are we doing threading?

File: data/delegation/store.go [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+	return nil
+}
+
+func (st *DelegationStore) UpdateWithdrawReward(height int64) {
+	st.mux.Lock()
+	defer st.mux.Unlock()
+
+	// st_m_ operation
+	fmt.Printf("--- Block mature update at height: %d ---\n", height)
```
Comment (alexjlan) [inline data/delegation/store.go:357]: can we put these log into detail log?

File: identity/validator_set.go [hunk 4]
```diff
@@ -137,23 +141,32 @@ func (vs *ValidatorStore) Setup(req types.RequestBeginBlock, nodeValidatorAddres
...
+	return def
+}
+
+func (vs *ValidatorStore) InitValidatorQueue(nodeValidatorAddress keys.Address) {
 	// initialize the queue for validators
 	vs.queue.PriorityQueue = make(utils.PriorityQueue, 0, 100)
 	i := 0
 	vs.totalPower = 0
 	vs.Iterate(func(addr keys.Address, validator *Validator) bool {
+		fmt.Printf("addr: %s - iterate!!!!!\n", addr.Humanize())
```
Comment (alexjlan) [inline identity/validator_set.go:155]: I think this is unnecessary log

**Pull request closed without merging**